src/audio/clap/plugin_manager.cpp

```diff
@@ -0,0 +1,309 @@
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+// The MSVC min/max macro imported by clap/all.h collides with
+// std::numeric_limits which makes the MSCV Debug x64 build fail.
+// Defining NOMINMAX is the workaround recommended by the MSCV team here:
+// https://github.com/microsoft/GSL/issues/816
+//
+#define NOMINMAX 1
+
+#include "plugin_manager.h"
+
+#include <cassert>
+#include <optional>
+
+#include "clap/all.h"
+
+#include "cross.h"
+#include "logging.h"
+#include "support.h"
+
+namespace Clap {
+
+static const clap_host_t dosbox_clap_host = {
+        .clap_version = CLAP_VERSION,
+
+        .host_data = nullptr,
+
+        .name    = "DOSBox Staging",
+        .vendor  = "The DOSBox Staging Team",
+        .url     = "http://www.dosbox-staging.org",
+        .version = "1.0",
+
+        .get_extension = []([[maybe_unused]] const clap_host_t* host,
+                            [[maybe_unused]] const char* extension_id) -> const void* {
+	        return nullptr;
+        },
+
+        .request_restart  = []([[maybe_unused]] const clap_host_t* host) {},
+        .request_process  = []([[maybe_unused]] const clap_host_t* host) {},
+        .request_callback = []([[maybe_unused]] const clap_host_t* host) {}};
+
+[[maybe_unused]] static std::optional<std_fs::path> find_first_file(const std_fs::path& path)
+{
+	std::vector<std_fs::path> files = {};
+
+	std::error_code ec;
+	for (const auto& entry : std_fs::directory_iterator(path, ec)) {
+		if (ec) {
+			return {};
+		}
+		if (entry.is_regular_file(ec)) {
+			files.emplace_back(entry.path());
+		}
+	}
+
+	if (files.size() == 0) {
+		return {};
+	}
+
+	std::sort(files.begin(), files.end());
+	return files.front();
+}
+
+static std::pair<dynlib_handle, const clap_plugin_entry_t*> load_plugin_library(
+        const std_fs::path& plugin_path)
+{
+	const auto reported_plugin_path = plugin_path;
+
+#ifdef MACOSX
+	// The dynamic-link library is usually inside the application bundle on
+	// macOS, so we need to resolve its path, but we must always report the
+	// bundle path to the plugin instance.
+	const auto library_path = [&] {
+		if (const auto f = find_first_file(plugin_path / "Contents" / "MacOS");
+		    f) {
+			return *f;
+		}
+		return plugin_path;
+	}();
+
+#else // Windows, Linux
+
+	// We use the path of the dynamic-link library directly on Windows and
+	// Linux.
+	const auto library_path = plugin_path;
+#endif
+
+	const auto lib = dynlib_open(library_path);
+	if (!lib) {
+		LOG_ERR("CLAP: Error loading plugin library '%s'",
+		        library_path.string().c_str());
+		return {};
+	}
+
+	const auto plugin_entry = reinterpret_cast<const clap_plugin_entry_t*>(
+	        dynlib_get_symbol(lib, "clap_entry"));
+
+	if (!plugin_entry) {
+		dynlib_close(lib);
+		return {};
+	}
+
+	plugin_entry->init(reported_plugin_path.string().c_str());
```

Comment:
Calling `plugin_entry->init()` more than once on a given library is also incorrect (again bad naming, this is actually a global init function for the library).  `mixer /listmidi` calls this on your SC-55 library 4 times

>  It also must only be called once, until a later call to deinit() is made, after which init() can be called once more to re-initialize the DSO.

Reply:
Ok, that's an oversight on my part. It kinda works now because I only end up using a single plugin instance (the first 3 of those 4 init calls are discovery and are paired with deinit/destroy calls). But it's incorrect, so I will do the clear plugin vs library separation as you suggested.

Reply:
The rickroll is paying off :laughing:   I started writing it just to get a grasp on how the CLAP API works and see if I  could catch any bugs in the host code.

Reply:
> The rickroll is paying off 😆 I started writing it just to get a grasp on how the CLAP API works and see if I could catch any bugs in the host code.

Nice man, that's dedication 🎖️